rle_array/_slicing.py

```diff
@@ -0,0 +1,395 @@
+            )
+
+        if self.start == self.stop:
+            # empty slice
+            if self.start != 0:
+                raise ValueError(
+                    f"for empty slices, start and stop must be 0 but are {self.start}"
+                )
+            if self.step != 1:
+                raise ValueError(f"for empty slices, step must be 1 but is {self.step}")
+        else:
+            # non-empty slice
+            if self.step == 0:
+                raise ValueError("step cannot be zero")
+            elif self.step > 0:
+                # forward slice
+                if self.start > self.stop:
+                    raise ValueError(
+                        "for forward slices, stop must be greater or equal to start"
+                    )
+            else:
+                # backward slice
+                if self.stop > self.start:
+                    raise ValueError(
+                        "for backward slices, start must be greater or equal to stop"
+                    )
+
+            if abs(self.start - self.stop) % abs(self.step) != 0:
+                raise ValueError("start->stop distance is not modulo step")
+
+    @property
+    def start(self) -> int:
+        """
+        Start index of the slice. Inclusive start.
+        """
+        return self._start
+
+    @property
+    def stop(self) -> int:
+        """
+        Stop index of the slice. Exclusive end.
+        """
+        return self._stop
+
+    @property
+    def step(self) -> int:
+        """
+        Step width.
+        """
+        return self._step
+
+    @property
+    def container_length(self) -> int:
+        """
+        Length of the container.
+        """
+        return self._container_length
+
+    def __repr__(self) -> str:
+        return (
+            f"{type(self).__name__}(start={self.start}, stop={self.stop}, step={self.step}, container_length="
+            f"{self.container_length})"
+        )
+
+    def __len__(self) -> int:
+        return self._calc_len(start=self.start, stop=self.stop, step=self.step)
+
+    @classmethod
+    def _calc_len(cls, start: int, stop: int, step: int) -> int:
+        """
+        Calculate slice length.
+
+        Parameters
+        ----------
+        start
+            Inclusive start index.
+        stop
+            Exclusive stop index.
+        step
+            Step width.
+        """
+        delta = abs(stop - start)
+        steps = delta // abs(step)
+        if delta % abs(step) != 0:
+            steps += 1
+        return steps
+
+    @classmethod
+    def _check_and_prepare_slice(cls, s: Optional[slice]) -> slice:
+        """
+        Check and prepare input slice for convertion.
+        """
+        if s is None:
+            s = slice(None, None, None)
+
+        if not isinstance(s, slice):
+            raise TypeError(f"slice must be a slice but is {type(s).__name__}")
+
+        if (s.start is not None) and not isinstance(s.start, (int, np.int64)):
```

Comment:
Do we actually have to check if it's an int or not when we already check it in the class init method?

Reply:
fail early